UefiPayloadPkg/Library/FdtParserLib/FdtParserLib.c

```diff
@@ -0,0 +1,698 @@
+          ASSERT (Serial != NULL);
+          if (Serial == NULL) {
+            break;
+          }
+
+          Serial->Header.Revision = UNIVERSAL_PAYLOAD_SERIAL_PORT_INFO_REVISION;
+          Serial->Header.Length   = sizeof (UNIVERSAL_PAYLOAD_SERIAL_PORT_INFO);
+          Serial->RegisterStride  = 1;
+
+          for (Property = FdtFirstPropertyOffset (Fdt, SSubNode); Property >= 0; Property = FdtNextPropertyOffset (Fdt, Property)) {
+            PropertyPtr = FdtGetPropertyByOffset (Fdt, Property, &TempLen);
+            TempStr     = FdtGetString (Fdt, Fdt32ToCpu (PropertyPtr->NameOffset), NULL);
+            if (AsciiStrCmp (TempStr, "current-speed") == 0) {
+              Data32 = (UINT32 *)(PropertyPtr->Data);
+              DEBUG ((DEBUG_INFO, "\n         Property(%08X)  %a", Property, TempStr));
+              DEBUG ((DEBUG_INFO, "  %X", Fdt32ToCpu (*Data32)));
+              Serial->BaudRate = Fdt32ToCpu (*Data32);
+            } else if (AsciiStrCmp (TempStr, "reg") == 0) {
+              Data32       = (UINT32 *)(PropertyPtr->Data);
+              Attribute    = Fdt32ToCpu (*(Data32 + 0));
+              StartAddress = Fdt32ToCpu (*(Data32 + 1));
+              DEBUG ((DEBUG_INFO, "\n in espi serial  Property(%08X)  %a", Property, TempStr));
+              DEBUG ((DEBUG_INFO, " StartAddress   %016lX\n", StartAddress));
+              DEBUG ((DEBUG_INFO, " Attribute      %016lX\n", Attribute));
+              Serial->RegisterBase = StartAddress;
+              Serial->UseMmio      = Attribute == 1 ? FALSE : TRUE;
+            }
+          }
+        }
+      }
+
+      for (Property = FdtFirstPropertyOffset (Fdt, Node); Property >= 0; Property = FdtNextPropertyOffset (Fdt, Property)) {
+        PropertyPtr = FdtGetPropertyByOffset (Fdt, Property, &TempLen);
+        TempStr     = FdtGetString (Fdt, Fdt32ToCpu (PropertyPtr->NameOffset), NULL);
+
+        if (AsciiStrCmp (TempStr, "ranges") == 0) {
+          DEBUG ((DEBUG_INFO, "  Found ranges Property TempLen (%08X)\n", TempLen));
+
+          PciRootBridgeInfo->RootBridge[index].AllocationAttributes = EFI_PCI_HOST_BRIDGE_COMBINE_MEM_PMEM | EFI_PCI_HOST_BRIDGE_MEM64_DECODE;
+          PciRootBridgeInfo->RootBridge[index].Supports             = ROOT_BRIDGE_SUPPORTS_DEFAULT;
+          PciRootBridgeInfo->RootBridge[index].PMemAbove4G.Base     = PcdGet64 (PcdPciReservedPMemAbove4GBBase);
+          PciRootBridgeInfo->RootBridge[index].PMemAbove4G.Limit    = PcdGet64 (PcdPciReservedPMemAbove4GBLimit);
+          PciRootBridgeInfo->RootBridge[index].PMem.Base            = PcdGet32 (PcdPciReservedPMemBase);
+          PciRootBridgeInfo->RootBridge[index].PMem.Limit           = PcdGet32 (PcdPciReservedPMemLimit);
+          PciRootBridgeInfo->RootBridge[index].UID                  = index;
+          PciRootBridgeInfo->RootBridge[index].HID                  = EISA_PNP_ID (0x0A03);
+
+          Data32 = (UINT32 *)(PropertyPtr->Data);
+          for (Base = 0; Base < 16; Base = Base + 7) {

```

Comment:
each range should be 6 elements per examples.
https://elinux.org/Device_Tree_Usage#PCI_Address_Translation
https://www.kernel.org/doc/Documentation/devicetree/bindings/pci/nvidia%2Ctegra20-pcie.txt

Reply:
sorry for confusion. In today meeting we just clarified the cpu_addr (parent address) in ranges can be either 1 cell or 2 cell, depending on address-cells in FDT root.
In our case we define root->address-cell as 2 so 2 cells for cpu_addr is correct. Please keep original implementation and no change needed here for now.

In future if we see 32bit cpu_addr support needed we can enhance parser to support it.